File: source/ellipse.cpp
#include "../stdafx.h"
#include "ellipse.h"

Ellipse::Ellipse(double radiusX, double radiusY)
{
    rx = radiusX;
    ry = radiusY;
}

double Ellipse::radius() const
{
    return std::max(rx, ry);
}

double Ellipse::step() const
{
    return 0.0; // не используем для эллипса
}

double Ellipse::length() const
{
    // Формула аппроксимации для периметра эллипса
    return M_PI * (3 * (rx + ry) - sqrt((3 * rx + ry) * (rx + 3 * ry)));
}

double Ellipse::curvature() const
{
    return 1.0 / std::max(rx, ry);
}

void Ellipse::evaluate(double t, double& x, double& y, double& z) const
{
    x = rx * cos(t);
    y = ry * sin(t);
    z = 0.0;
}

void Ellipse::derivative(double t, double& dx, double& dy, double& dz) const
{
    dx = -rx * sin(t);
    dy = ry * cos(t);
    dz = 0.0;
}

File: internal/ellipse/ellipse.go
package ellipse

import (
	"math"
)

type Ellipse struct {
	rx, ry float64
}

func NewEllipse(radiusX, radiusY float64) *Ellipse {
	return &Ellipse{rx: radiusX, ry: radiusY}
}

func (e *Ellipse) Radius() float64 {
	return math.Max(e.rx, e.ry)
}

func (e *Ellipse) Step() float64 {
	return 0.0
}

func (e *Ellipse) Length() float64 {
	return math.Pi * (3*(e.rx+e.ry) - math.Sqrt((3*e.rx+e.ry)*(e.rx+3*e.ry)))
}

func (e *Ellipse) Curvature() float64 {
	return 1.0 / math.Max(e.rx, e.ry)
}

func (e *Ellipse) Evaluate(t float64) (x, y, z float64) {
	x = e.rx * math.Cos(t)
	y = e.ry * math.Sin(t)
	z = 0.0
	return
}

func (e *Ellipse) Derivative(t float64) (dx, dy, dz float64) {
	dx = -e.rx * math.Sin(t)
	dy = e.ry * math.Cos(t)
	dz = 0.0
	return
}
